### ISSUE TITLE
Including version number in module path breaks Go modules

### ISSUE THREAD

**Issue by twpayne** [google/go-github#1199]
Including the version number, e.g. `v26`, in the module path breaks Go modules as `v25` was deleted.

See https://github.com/twpayne/chezmoi/pull/355 for an example of this breaking a project.

Please consider removing the version number from the path and using Go modules as Russ intended :)

**Comment by gmlewis** [comment]
Sorry, you caught me a little off-guard...  First off, why do you say "`v25` was deleted"? The whole point of `go mod` is to find a version (in this case, on GitHub, where *nothing* is deleted) and allow you to stick to it as a dependency of your project.

Second, you say to use "Go modules as Russ indended"... can you please follow that statement with some documentation pointing out how we are *not* doing that?

Looking through our code base, the version number is only referenced in these files:

```
README.md
go.mod
test/integration/users_test.go
test/integration/repos_test.go
test/fields/fields.go
test/integration/activity_test.go
test/integration/authorizations_test.go
test/integration/github_test.go
github/doc.go
github/examples_test.go
example/basicauth/main.go
example/simple/main.go
example/newrepo/main.go
example/appengine/app.go
example/migrations/main.go
example/commitpr/main.go
```

...none of which should break anything.

Additionally, our first usage of versioning started with `v18` about a year ago, and this is the very first report of us doing versioning differently from how "Russ intended". Russ actually comments on this repo every now and then and helps out when we have questions, so I would be really surprised that we haven't been corrected after a year of operating this way. We also have thousands of users and I believe at least one of them is using versioning. Maybe you are not doing versioning the way "Russ intended"?

So, can you please give a little more details than you have provided?  All I see in that link is someone bumping their repo to point to `v26`, which is not a very convincing argument that we broke something or that we are doing versioning incorrectly.

Thanks.

**Comment by twpayne** [comment]
Sure. The problem is that the version number is in the package name, whereas it should not be, and instead it should be a git tag.

Say a project is using `github.com/google/go-github/v25`, with Go modules, the `go` tool goes looking in https://github.com/google/go-github for a module with that name. The version has now been bumped and the module name has been changed to`github.com/google/go-github/v26`. Consequently, the module `github.com/google/go-github/v25` no longer exists.

See [the Go wiki page on modules](https://github.com/golang/go/wiki/Modules), and particularly [publishing a release](https://github.com/golang/go/wiki/Modules#publishing-a-release).

tl;dr:

* Remove the version number from the name of the module.
* Use git tags to tag versions.

**Comment by gmlewis** [comment]
So if I understand correctly, `/v26` needs to be removed from the first line of `go.mod`, correct?

Every time we make a "release", GitHub tags the release for us: https://github.com/google/go-github/releases
Aren't these sufficient?

**Comment by twpayne** [comment]
> So if I understand correctly, `/v26` needs to be removed from the first line of go.mod, correct?

Yes, this is my understanding too.

GitHub doesn't tag releases for you, the tag is created by you (or maybe by some release automation software if you're using any). When a tag gets pushed to GitHub, GitHub will create an entry on the releases page for you, but to the best of my knowledge these are not used by Go in any way.

**Comment by gmlewis** [comment]
OK, right... I manually create the tag name for each release.

So I still don't understand your `Use git tags to tag version` comment. Are you saying more modifications are needed for `go.mod`?

Do you want to show me how `go.mod` should look or send me a PR, or is the deletion of `/v26` sufficient?

**Comment by twpayne** [comment]
The deletion the `/v26` should be sufficient. You'll need to delete it everywhere:

```
grep -rl /v26 | xargs sed -i s:/v26::g
```

As this will change the module name for users, I'd also tag a new version `v27.0.0` after deleting `/v26`.

**Comment by twpayne** [comment]
To tag a new version:

```
git tag v27.0.0
git push origin --tags
```

**Comment by gmlewis** [comment]
OK, but this statement ("delete it everywhere") I truly do not understand... please take a look at the list of filenames where it is used. None of the files are in the repo source code itself except in examples and integration tests.

**Comment by twpayne** [comment]
Deleting `/v26` from the module name will change the module name. So, anything like:

```go
import "github.com/google/go-github/v26/github"
```

has to be changed to

```go
import "github.com/google/go-github/github"
```

This includes tests, examples, etc.

**Comment by gmlewis** [comment]
@dsymonds - if you have a few minutes, could you please take a look at this issue and comment?

**Comment by gmlewis** [comment]
Also, maybe you have a few minutes to comment, @campoy?

**Comment by twpayne** [comment]
It's possible that I've made a mistake here. It looks like `github.com/google/go-github` is following the [releasing modules (v2 or higher) strategy described in the Wiki](https://github.com/golang/go/wiki/Modules#releasing-modules-v2-or-higher), and it's possible that the [triggering issue](https://github.com/twpayne/chezmoi/pull/355) hit problems because the user's Go version does not support this. I'll ask.

**Comment by campoy** [comment]
Oof, Go modules ain't easy at scale if you ask me.

I think the one that can help here is @broady

**Comment by twpayne** [comment]
I think I messed up here. Big apologies for the noise and distraction. Please re-open if I missed something.

**Comment by gmlewis** [comment]
No worries, @twpayne - we want this to be the most useful package it can be... so if you do find problems definitely let us know. Thanks.
